crates/matrix-sdk/src/client/builder.rs

```diff
@@ -641,3 +738,273 @@ impl ClientBuildError {
+        let homeserver = make_mock_homeserver().await;
+        let mut builder = ClientBuilder::new();
+
+        Mock::given(method("GET"))
+            .and(path("/.well-known/matrix/client"))
+            .respond_with(ResponseTemplate::new(200).set_body_json(make_well_known_json(
+                &homeserver.uri(),
+                Some("https://localhost:1234"),
+                Some("https://localhost:5678"),
+            )))
+            .mount(&server)
+            .await;
+
+        // When building a client with the base server.
+        builder = builder.server_name_or_homeserver_url(server.uri());
+        let client = builder.build().await.unwrap();
+
+        // Then a client should be built with support for both sliding sync and OIDC.
+        #[cfg(feature = "experimental-sliding-sync")]
+        assert_eq!(client.sliding_sync_proxy(), Some("https://localhost:1234".parse().unwrap()));
+        #[cfg(feature = "experimental-oidc")]
+        assert_eq!(
+            client.oidc().authentication_server_info().unwrap().issuer,
+            "https://localhost:5678".to_owned()
+        );
+    }
+
+    /* Helper functions */
+
+    async fn make_mock_homeserver() -> MockServer {
+        let homeserver = MockServer::start().await;
+        Mock::given(method("GET"))
+            .and(path("/_matrix/client/versions"))
+            .respond_with(ResponseTemplate::new(200).set_body_json(&*test_json::VERSIONS))
+            .mount(&homeserver)
+            .await;
+        Mock::given(method("GET"))
+            .and(path("/_matrix/client/r0/login"))
+            .respond_with(ResponseTemplate::new(200).set_body_json(&*test_json::LOGIN_TYPES))
+            .mount(&homeserver)
+            .await;
+        homeserver
+    }
+
+    fn make_well_known_json(
+        homeserver_url: &str,
+        sliding_sync_proxy_url: Option<&str>,
+        authentication_issuer: Option<&str>,
+    ) -> JsonValue {
+        ::serde_json::Value::Object({
+            let mut object = ::serde_json::Map::new();
+            let _ = object.insert(
+                "m.homeserver".into(),
+                json_internal!({
+                    "base_url": homeserver_url
+                }),
+            );
+
+            if let Some(sliding_sync_proxy_url) = sliding_sync_proxy_url {
+                let _ = object.insert(
+                    "org.matrix.msc3575.proxy".into(),
+                    json_internal!({
+                        "url": sliding_sync_proxy_url
+                    }),
+                );
+            }
+
+            if let Some(authentication_issuer) = authentication_issuer {
+                let _ = object.insert(
+                    "org.matrix.msc2965.authentication".into(),
+                    json_internal!({
+                        "issuer": authentication_issuer,
+                        "account": authentication_issuer.to_owned() + "/account"
+                    }),
+                );
+            }
+
+            object
```

Comment:
I'm wondering if this could be

```rust
json!({
  "m.homeserver": {
    "base_url": homeserver_url
  },
  "org.matrix.msc3575.proxy": {
    "url": sliding_sync_proxy_url // will be null when the option is set to none
  }, // etc
})
```

Reply:
I have a feeling it will fail as I think url is a required property so presumably not an `Option`, but I'll give it a go.

Reply:
Oh you might be right; in that case, we can keep the code as is.

Reply:
Yeah, sadly this doesn't work, as it results in deserialization errors on `null` values.